crates/bevy_pbr/src/deferred/mod.rs

```diff
@@ -0,0 +1,451 @@
+use bevy_app::prelude::*;
+use bevy_asset::{load_internal_asset, Handle};
+use bevy_core_pipeline::{
+    clear_color::ClearColorConfig,
+    core_3d,
+    deferred::{
+        copy_lighting_id::DeferredLightingIdDepthTexture, DEFERRED_LIGHTING_PASS_ID_DEPTH_FORMAT,
+    },
+    prelude::{Camera3d, ClearColor},
+    prepass::DeferredPrepass,
+    tonemapping::{DebandDither, Tonemapping},
+};
+use bevy_ecs::{prelude::*, query::QueryItem};
+use bevy_render::{
+    extract_component::{
+        ComponentUniforms, ExtractComponent, ExtractComponentPlugin, UniformComponentPlugin,
+    },
+    render_asset::RenderAssets,
+    render_graph::{NodeRunError, RenderGraphContext, ViewNode, ViewNodeRunner},
+    render_resource::{self, Operations, PipelineCache, RenderPassDescriptor},
+    renderer::{RenderContext, RenderDevice},
+    texture::Image,
+    view::{ViewTarget, ViewUniformOffset},
+    Render, RenderSet,
+};
+
+use bevy_render::{
+    render_graph::RenderGraphApp, render_resource::*, texture::BevyDefault, view::ExtractedView,
+    RenderApp,
+};
+
+use crate::{
+    EnvironmentMapLight, MeshPipelineKey, ViewFogUniformOffset, ViewLightsUniformOffset,
+    MAX_CASCADES_PER_LIGHT, MAX_DIRECTIONAL_LIGHTS,
+};
+
+pub struct PbrDeferredLightingPlugin;
+
+pub const DEFERRED_LIGHTING_SHADER_HANDLE: Handle<Shader> =
+    Handle::weak_from_u128(2708011359337029741);
+
+pub const DEFAULT_PBR_DEFERRED_LIGHTING_DEPTH_ID: u8 = 1;
+
+/// Component with a `depth_id` for specifying which corresponding materials should be rendered by this specific PBR deferred lighting pass.
+/// Will be automatically added to entities with the [`DeferredPrepass`] component that don't already have a [`PbrDeferredLightingDepthId`].
+#[derive(Component, Clone, Copy, ExtractComponent, ShaderType)]
+pub struct PbrDeferredLightingDepthId {
+    depth_id: u32,
+
+    #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+    _webgl2_padding: bevy_math::Vec3,
+}
+
+impl PbrDeferredLightingDepthId {
+    pub fn new(value: u8) -> PbrDeferredLightingDepthId {
+        PbrDeferredLightingDepthId {
+            depth_id: value as u32,
+
+            #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+            _webgl2_padding: bevy_math::Vec3::ZERO,
+        }
+    }
+
+    pub fn set(&mut self, value: u8) {
+        self.depth_id = value as u32;
+    }
+
+    pub fn get(&self) -> u8 {
+        self.depth_id as u8
+    }
+}
+
+impl Default for PbrDeferredLightingDepthId {
+    fn default() -> Self {
+        PbrDeferredLightingDepthId {
+            depth_id: 1,
+
+            #[cfg(all(feature = "webgl", target_arch = "wasm32"))]
+            _webgl2_padding: bevy_math::Vec3::ZERO,
+        }
+    }
+}
+
+impl Plugin for PbrDeferredLightingPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins((
+            ExtractComponentPlugin::<PbrDeferredLightingDepthId>::default(),
+            UniformComponentPlugin::<PbrDeferredLightingDepthId>::default(),
```

Comment:
Would these be better in an array in a binding? And then indexing into them.

Reply:
Sorry, I don't follow.

Reply:
`UniformComponentPlugin` requires a dynamic offset binding for each individual value, which breaks batching. I was asking whether it would make more sense to put all values into an array in a binding, like with `GpuArrayBuffer`, and then index into the array by passing in the index somehow. I'm guessing it's not that important though?

Reply:
This is for the `PbrDeferredLightingPlugin` which involves an entire render pass of a full screen triangle. No use batching here.